Ensure ShardedTensor handles list/tuple appropriately as `size` parameter.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62109

The `size` parameter only worked correctly for *args like invocation
:10, 20 and not for list: [10, 20] and tuples: (10, 20). This PR ensures this
works similar to `torch.empty`.

Differential Revision: [D29884768](https://our.internmc.facebook.com/intern/diff/D29884768/)

pritamdamania committed:
Ensure ShardedTensor handles list/tuple appropriately as `size` parameter.

The `size` parameter only worked correctly for *args like invocation
:10, 20 and not for list: [10, 20] and tuples: (10, 20). This PR ensures this
works similar to `torch.empty`.

Differential Revision: [D29884768](https://our.internmc.facebook.com/intern/diff/D29884768/)

[ghstack-poisoned]

SciPioneer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -472,6 +472,32 @@ def test_insufficient_sharding_dims(self):
+            ],
+        )
+
+        # Test with *args
```
[Optional] Nit: you can loop over `_sharded_tensor.empty(spec, ...)` to reduce the redundant code `self.assertEqual(torch.Size([10, 20]), sharded_tensor.size())`. It probably won't save too much though, so totally up to you.

pritamdamania87 replied:
Not sure how this would work concisely, the loop somehow would need to iterate of things like 10, 20, [10, 20] and (10, 20) separately.

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -136,7 +136,7 @@ def __init__(
             raise ValueError('Only torch.contiguous_format memory_format is currently supported')
 
-        dims = list(size)
+        dims = list(*size) if len(size) == 1 else list(size)
```
why `len(size) == 1` we still unpack the `size`? it seems to me when length==1, size will just become a tuple like `(3,)`, shouldn't this still work?

pritamdamania87 replied:
The problem is when size is a list, what list(size) does is it creates a List[List[]]

pritamdamania committed:
Update on "Ensure ShardedTensor handles list/tuple appropriately as `size` parameter."


The `size` parameter only worked correctly for *args like invocation
:10, 20 and not for list: [10, 20] and tuples: (10, 20). This PR ensures this
works similar to `torch.empty`.

Differential Revision: [D29884768](https://our.internmc.facebook.com/intern/diff/D29884768/)

[ghstack-poisoned]
